taichi/analysis/same_statements.cpp

```diff
@@ -0,0 +1,234 @@
+#include "taichi/ir/ir.h"
+#include <map>
+#include <set>
+
+TLANG_NAMESPACE_BEGIN
+
+#define DEFINE_TYPE_CHECK(Type)       \
+  if (!other_node->is<Type>()) {      \
+    same = false;                     \
+    return;                           \
+  }                                   \
+  auto other = other_node->as<Type>();
+
+#define DEFINE_OPERAND_CHECK                                           \
+  if (stmt->num_operands() != other->num_operands()) {                 \
+    same = false;                                                      \
+    return;                                                            \
+  }                                                                    \
+  for (int i = 0; i < stmt->num_operands(); i++) {                     \
+    if (get_other_id(stmt->Stmt::operand(i)->id)                       \
+        != other->Stmt::operand(i)->id) {                              \
+      /*to distinguish from UnaryOpStmt::operand*/                     \
+      same = false;                                                    \
+      return;                                                          \
+    }                                                                  \
+  }
+
+#define DEFINE_FIELD_CHECK(field)    \
+  if (stmt->field != other->field) { \
+    same = false;                    \
+    return;                          \
+  }
+
+#define DEFINE_MAP_ID map_id(stmt->id, other->id);
+
+#define DEFINE_BASIC_CHECK(Type) \
+  DEFINE_TYPE_CHECK(Type)        \
+  DEFINE_OPERAND_CHECK           \
+  DEFINE_MAP_ID
+
+// Compare if two IRNodes are equivalent.
+class IRNodeComparator : public IRVisitor {
+ private:
+  IRNode *other_node;
+  std::map<int, int> id_map; // map the id from this node to the other node
+  std::set<int> captured_id; // ids which don't belong to either node
```

Comment:
We may need to use the `std::unordered_` versions for efficienty.

Reply:
What's the advantage of map? If unordered_map always superior, then why they don't replace it then?

Reply:
`std::map` uses less memory, provides functions about the order (`lower_bound` for example), and guarantees worst-case O(log(n)) time (`std::unordered_map` has a worst-case complexity of O(n)). I agree that we should use `std::unordered_map` in this case.

Reply:
I just noticed that the indices inserted to `id_map` must be in ascending order. Maybe `std::vector` is more efficient if the range is not too big (and after `re_id`, it is not). However, I would still prefer to use `std::unordered_map` for maintainable code.